Fix #609 - Hoodie default page not showing

minrwhite opened the pull request:
closes #609

minrwhite committed:
test(cli/options): check that webroot locator module consumed

Add requirement for a webroot-locator module with the assumption that
this will be resolving the correct webroot to use based on directory
existence

minrwhite committed:
fix(cli/options): ensure that options-test passes

Pass the options-test.js test added previously by requiring
webroot-locator.js and using its result. Note that
webroot-locator.js does not exist, it is to be stubbed for the
purposes of the test.

minrwhite committed:
test(cli/webroot-locator): test webroot locator functionality

Test the functionality to be added in webroot-locator module; namely
that if a path is passed in and it exists and is a directory, then
the the module shall return the path verbatim, otherwise the module
will resolve and return the fallback public path.

minrwhite committed:
fix(cli/webroot-locator): ensure that test passes

Create webroot-locator.js such that the previously created test
passes, additionally resolving any errors with the test itself.

Closes hoodiehq/hoodie#609

gr2m commented on cli/webroot-locator.js:
```diff
@@ -0,0 +1,11 @@
+module.exports = locateWebroot
+var fs = require('fs')
```
🎨 can you please add an empty line below `module.exports = locateWebroot`

gr2m commented on test/unit/cli/webroot-locator-test.js:
```diff
@@ -0,0 +1,73 @@
+    {fs: fsMock})
+
+  group.afterEach(function (done) {
+    fsMock.existsSync.actions.length = 0
```
what is the `.actions.length = 0` for?

minrwhite replied:
Comment moved from below: It's shorthand for emptying an array. I can change this to be a bit more explicit if required?

gr2m replied:
okay I understand what `stub.actions` are now. I didn’t know one can chain `callbackWith`, `returnWith` and `throwWith`, that’s pretty cool! I would think that `stub.reset()` would remove these though, I’ll file an issue on simple-mock.

But please change the `actions.length = 0` to `actions = []`, it’s shorter and more readable

gr2m commented on test/unit/cli/options-test.js:
```diff
@@ -36,6 +36,8 @@ var packageJsonMock = {
   version: '1.0.0'
 }
 
+var mockWebrootLocator = simple.spy(function (arg) { return arg })
```
I didn’t know that you can stub `simple.spy`s, too, but for coherence with other code, can you do `simple.stub()` here and in the other places please?

gr2m reviewed (changes requested):
looking great, thanks! Just a few comments

minrwhite committed:
fix(cli/options): Review changes

minrwhite committed:
test(plugin/public): Test correct public config path

minrwhite commented:
actions.length = 0 is a way of resetting the stubbed behaviour, as spy.reset() only resets the logged calls. That way if we forget to change the behaviour between tests this should catch it and prevent a false positive.

Currently this fails, as public plugin currently determines the public webroot from the wrong config object path. Correcting this causes other tests to fail; I'll fix these tests as my next course of action.

gr2m commented:
but you don’t use `actions.length` anywhere else in the code? Am I missing something?

minrwhite commented:
It's shorthand for emptying an array. I can change this to be a bit more explicit if required?

minrwhite committed:
fix(plugin/public): Correct object path to public config

minrwhite committed:
fix(test/integration): Fix integration test failures

Following the previous change, integration tests were failing due
to a missing object on the passed config. Add the missing object
to the stubbed config in order to prevent these failures.

Closes #609

minrwhite commented:
Rebased and pushed, ready for review, also pending yay/nay on the `actions.length = 0` array-emptying shorthand.

gr2m commented on test/unit/cli/options-test.js:
```diff
@@ -36,6 +36,8 @@ var packageJsonMock = {
   version: '1.0.0'
 }
 
+var mockWebrootLocator = simple.stub(function (arg) { return arg })
```
you can remove `function (arg) { return arg }`

minrwhite replied:
Indeed I can! Done.

gr2m commented on test/unit/cli/webroot-locator-test.js:
```diff
@@ -0,0 +1,73 @@
+
+test('webroot locator', function (group) {
+  var webrootLocator = proxyquire('../../../cli/webroot-locator',
+    {fs: fsMock})
```
can you please change this to 

``` js
var webrootLocator = proxyquire('../../../cli/webroot-locator', {
  fs: fsMock
})
```

minrwhite replied:
Changed. Agreed that that's clearer.

gr2m reviewed (changes requested):
okay final comments. I can also fix it up myself if you are busy?

minrwhite committed:
test: Changes requested for review

minrwhite commented:
Actioned review changes, and rebased ready for review.

gr2m commented:
beautiful work, thanks a ton

brunopedroso commented:
ooops.. should we check if the `server.paths` key is present? otherwise I'm forced to define it like

```
paths:{}
```

I'm using ember-hoodie...

gr2m commented:
it gets defaulted here: https://github.com/hoodiehq/hoodie/blob/047a8ff004538e68ccf2a89df33108fec505c3f3/server/index.js#L35-L40

brunopedroso commented:
yes, but only if I [dont define](https://github.com/hoodiehq/hoodie/blob/047a8ff004538e68ccf2a89df33108fec505c3f3/server/index.js#L26) `db.url`.

We're setting `db.url` here, and our app started to break on [this line](https://github.com/hoodiehq/hoodie/pull/627/files#diff-43557d6060003ff6997c94943ed23a82R11) just after I made the update. I had to define `paths:{}`...

Maybe that line could read like this?

```
var publicFolder = (options.config.paths && options.config.paths.public) || 'public'
```

minrwhite commented:
FWIW, I'd be tempted to resolve this by moving [this clause](/hoodiehq/hoodie/blob/047a8ff004538e68ccf2a89df33108fec505c3f3/server/index.js#L35) alongside [this one](/hoodiehq/hoodie/blob/047a8ff004538e68ccf2a89df33108fec505c3f3/server/index.js#L19), removing the 'data' key and letting [this](/hoodiehq/hoodie/blob/047a8ff004538e68ccf2a89df33108fec505c3f3/server/index.js#L41) handle the data key.

gr2m commented:
@brunopedroso ooops that is definitely wrong

The way it’s currently handled is not very elegant. I’ve a lot of work stuff to catch up with and won’t be able to look into this, but if one of you could send a PR that will fix @brunopedroso’s problem, I’m happy to review and merge
